### DISCUSSION TITLE
A solution for slower PWM scenarios

### DISCUSSION THREAD

**Discussion by seanmcveigh** [arendst/Tasmota#18242] [Q&A]
Hi,
Long time fan, first time hacker on this codebase :)
I've got an ESP8266 controlling a heating element on a zero-crossing SSR with a response time around 10ms so kilohertz PWM frequencies are not usable in this situation.  A PWM frequency in the <5Hz range would be ideal - even as low as 0.1Hz depending on the power resolution required.
I looked into maybe some rules/timer based duty cycling with the gpio configured for relay service.  This is sort of usable, but the hurdles I've encountered are:

- when using something like `ON Power1#state=0 DO Backlog Delay 1; Power1 1 ENDON ON Power1#state=1 DO Backlog Delay 3; Power1 0 ENDON`, the minimum pulse timing ends up being somewhere between 200-300ms due to Backlog's inter-command delays.  I haven't tried dialing that delay down. using Backlog0 in place causes the Delay to be ignored.
- looked at using RuleTimer, but the resolution is only 1 second which could be usable in conjunction with PulseTime, but this ends up limiting the granularity (ie. 4% duty cycle would require 200ms on time per 5 seconds instead of 100ms per 2.5 seconds, which is a bit smoother).
- looked at BlinkTime as a solution, but this has a minimum of 200ms and only 50% duty cycle. This is also a global configuration.
- the PWM code as-is seems to be married to a hardware implementation. It would be fairly simple to write a software-based PWM solution similar to the blink handling, but would need a new Scheduler hook, and not sure if the command API would easily support lower frequencies without some adjustment (e.g., for the sub-1-Hz range).

As a test, I went and tweaked the blink code path by doing the following:

- add a new command BlinkOffTime which takes a negative duty cycle time period (0.1-360s)
- adjusted BlinkTime to allow 0.1 second rather than the current minimum 0.2 seconds (is there a reason for this being the current floor value?)
- in the blink timer hook, I applied the BlinkTime timer period when turning the relay from off to on, and apply BlinkOffTime when turning the relay from on to off.  This gives an asymmetric blink pattern.

This works well for my purposes and I can get 1% thru 99% duty cycle to fit in a variable time period ranging from 0.1-10 seconds (worst case for 1% and 99%) with most values running faster of course - like 50% in 0.2 seconds or 33% in 0.3 seconds.

One thing this solution lacks of course is a nice way to program the duty cycle - using a slider or otherwise.  For now, I'm probably going to just configure up a dimmer on an unused gpio and trigger some rules off of changes to that slider that will do a bit of math and update BlinkTime and BlinkOffTime appropriately.

If anyone is interested in this sort of thing, I can share my patch which is only about 25 lines.

If anyone would like to see some more flexible PWM support for lower frequency devices like this using the built-in PWM codebase, I could take a look at splitting off a software-vs-hardware implementation.  As I mentioned, I'm new to the codebase though, so not sure if Theo or others would have some feedback on that plan? (e.g., additional overhead in the sub-second scheduler hooks)

Anyhow.. good times with Tasmota!

**Comment by barbudor** [comment]
Have you looked into the Thermostat feature?
It looks like designed for what you want to achieve

Alternatively you can use Pulsetime to fix a duration and use RuleTimer to trigger the power on at a variable period. So actually changing the duty cycle by changing the period with a fixed on time.
I'm using a similar method with only rule to drive an electric heater. Pulsetile is set to 3'30 and every 5' I get a room temp from my BLE sensor and I decide to power on or not. Works fine and the room is regulated within 0.5°C

**Reply by seanmcveigh** [reply]
Thanks for the response!  I don't believe the thermostat feature will help me, since I am not trying to reach a set-point in a closed-loop fashion. I am just trying to vary power output in a granular way and without cycling between max power and min power (on/off) over long periods of time.  As I mentioned, I did look at Pulsetime and RuleTimer, but the resolution are not what I am after.

I actually just went ahead and did a second implementation over the past hour that provides arbitrary duty-cycle blink functionality using a cumulative-error-reducing algorithm similar to Bresenham.  About 40 lines of code.  Command to configure is `BlinkDuty 33` (33% for example).  Looks to be cycling on for 100ms, off for 200ms and repeating.

**Reply by sfromis** [reply]
Another option could be to use ESP32, where `PWMFrequency` is down to 2 Hz.
